sdk/include/opentelemetry/sdk/metrics/state/metric_collector.h

```diff
@@ -0,0 +1,57 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#ifndef ENABLE_METRICS_PREVIEW
+#  include "opentelemetry/sdk/metrics/data/metric_data.h"
+#  include "opentelemetry/sdk/metrics/export/metric_producer.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace metrics
+{
+
+class MetricReader;
+class MeterContext;
+
+class CollectorHandle
+{
+public:
+  virtual AggregationTemporarily GetAggregationTemporarily() noexcept = 0;
+};
+
+/**
+ * An internal opaque interface that the MetricReader receives as
+ * MetricProducer. It acts as the storage key to the internal metric stream
+ * state for each MetricReader.
+ */
+
+class MetricCollector : public MetricProducer, public CollectorHandle
+{
+public:
+  MetricCollector(std::shared_ptr<MeterContext> &&context,
```

Comment:
Why do we need a `shared_ptr` for context?

Reply:
It would be used across all the meters created, so it is maintained as shared_ptr in the [meter provider](https://github.com/open-telemetry/opentelemetry-cpp/blob/49648bd8f741a28a2b111c8d8ecad75a070ea649/sdk/include/opentelemetry/sdk/metrics/meter_provider.h#L106).

Reply:
Thanks for the clarification.